src/daft-schema/src/schema.rs

```diff
@@ -10,99 +9,128 @@ use common_display::{
+    fields: Vec<Field>,
+
+    #[derivative(Hash = "ignore", PartialEq = "ignore")]
+    name_to_indices: HashMap<String, Vec<usize>>,
 }
 
 impl Schema {
-    pub fn new(fields: Vec<Field>) -> DaftResult<Self> {
-        let mut map = IndexMap::new();
-
-        for f in fields {
-            match map.entry(f.name.clone()) {
-                indexmap::map::Entry::Vacant(entry) => {
-                    entry.insert(f);
-                }
-                indexmap::map::Entry::Occupied(entry) => {
-                    return Err(DaftError::ValueError(format!(
-                        "Attempting to make a Schema with duplicate field names: {}",
-                        entry.key()
-                    )));
-                }
+    pub fn new(fields: Vec<Field>) -> Self {
+        let mut name_to_indices = HashMap::<String, Vec<usize>>::new();
+
+        for (i, f) in fields.iter().enumerate() {
+            if let Some(indices) = name_to_indices.get_mut(&f.name) {
+                indices.push(i);
+            } else {
+                name_to_indices.insert(f.name.clone(), vec![i]);
             }
         }
 
-        Ok(Self { fields: map })
+        Self {
+            fields,
+            name_to_indices,
+        }
+    }
+
+    pub fn empty() -> Self {
+        Self::new(vec![])
     }
 
     pub fn to_struct(&self) -> DataType {
-        let fields = self.fields.values().cloned().collect();
-        DataType::Struct(fields)
+        DataType::Struct(self.fields.clone())
     }
 
-    pub fn exclude<S: AsRef<str>>(&self, names: &[S]) -> DaftResult<Self> {
-        let mut fields = IndexMap::new();
-        let names = names.iter().map(|s| s.as_ref()).collect::<HashSet<&str>>();
-        for (name, field) in &self.fields {
-            if !names.contains(&name.as_str()) {
-                fields.insert(name.clone(), field.clone());
-            }
-        }
+    pub fn get_field_at_index(&self, index: usize) -> DaftResult<&Field> {
+        self.fields
+            .get(index)
+            .ok_or(DaftError::FieldNotFound(format!(
+                "Attempted to access field at out-of-bounds index {} in schema: {:?}",
+                index, self.fields
+            )))
+    }
 
-        Ok(Self { fields })
+    pub fn fields(&self) -> &[Field] {
+        &self.fields
     }
 
-    pub fn empty() -> Self {
-        Self {
-            fields: indexmap::IndexMap::new(),
-        }
+    pub fn field_names(&self) -> impl Iterator<Item = &str> {
+        self.fields.iter().map(|f| f.name.as_str())
+    }
+
+    #[deprecated(since = "TBD", note = "name-referenced columns")]
+    pub fn exclude<S: AsRef<str>>(&self, names: &[S]) -> Self {
+        let names = names.iter().map(|s| s.as_ref()).collect::<HashSet<&str>>();
+        let fields = self
+            .fields
+            .iter()
+            .filter(|field| !names.contains(field.name.as_str()))
+            .cloned()
+            .collect::<Vec<_>>();
+
+        Self::new(fields)
```

Comment:
Perhaps this becomes our filter impl 😁

Reply:
For sure, replace `names` with a predicate

Reply:
Sounds good. I think I will also leave that to a future PR when it is needed